contracts/DelegateContribute.sol

```diff
@@ -0,0 +1,120 @@
+    /// @param delegatee provided delegatee.
+    event Delegation(address indexed delegator, address indexed delegatee);
+
+    // wveOLAS address
+    address immutable wveOLAS;
+
+    // Mapping of account address => address
+    mapping(address => address) public mapDelegation;
+
+    // Maps a delegatee to their list of delegators
+    mapping(address => address[]) private delegatorLists;
+
+    // Maps a delegator to their index in the delegatee's list + 1 (since default value is 0)
+    mapping(address => mapping(address => uint256)) private delegatorIndex;
+
+    /// @dev DynamicContribution constructor.
+    /// @param _wveOLAS wveOLAS contract address.
+    constructor(address _wveOLAS)
+    {
+        wveOLAS = _wveOLAS;
+    }
+
+    /// @dev Delegates Contribute voting power to an address.
+    /// @param delegatee Delegatee address.
+    function delegate(address delegatee) external {
+        address delegator = msg.sender;
+        if (delegator == delegatee) {
+            revert NoSelfDelegation(delegator);
+        }
+        address currentDelegatee = mapDelegation[delegator];
+        if (currentDelegatee == delegatee) {
+            revert AlreadyDelegatedToSameDelegatee(delegator, delegatee);
+        }
+        uint256 balanceOf = IWVEOLAS(wveOLAS).balanceOf(delegator);
+        if (balanceOf == 0) {
+            revert NoBalance(delegator);
+        }
+
+        // Remove from old delegatee if applicable
+        if(currentDelegatee != address(0)) {
+            uint256 index = delegatorIndex[currentDelegatee][delegator];
+
+            // Adjust index since it's stored +1
+            index -= 1;
+
+            // Move the last element to the deleted spot to maintain compact array
+            address lastDelegator = delegatorLists[currentDelegatee][delegatorLists[currentDelegatee].length - 1];
+            delegatorLists[currentDelegatee][index] = lastDelegator;
+            delegatorIndex[currentDelegatee][lastDelegator] = index + 1; // Update moved delegator's index
+
+            // Remove the last element
+            delegatorLists[currentDelegatee].pop();
+            delete delegatorIndex[currentDelegatee][delegator];
+        }
+
+        mapDelegation[delegator] = delegatee;
+
+        if(delegatee != address(0)) {
+            delegatorLists[delegatee].push(delegator);
+            // Store the index of the new delegator, +1 to differentiate from default value
+            delegatorIndex[delegatee][delegator] = delegatorLists[delegatee].length;
+        }
+
+        emit Delegation(delegator, delegatee);
+    }
+
+    /// @dev Checks for the votingPower of a delegatee.
+    /// @notice the votingPower includes the delegatee's own voting power, unless delegated.
+    /// @param delegatee Delegatee address.
+    /// @return totalVotingPower total voting power of delegatee.
+    function votingPower(address delegatee) external view returns (uint256 totalVotingPower) {
```

Comment:
See note above. (*)
If the function will not be used from another contract, but will be called from code, then I do not see any problems. Otherwise, a delegators.length limitation must be introduced. Otherwise the cycle will not complete due to lack of gas.